File: examples/nodejs/vault-secrets.js
/**
 * HashiCorp Vault Integration
 *
 * WHY VAULT OVER .env FILES:
 * ───────────────────────────
 * Vault provides dynamic secrets (short-lived credentials generated on demand),
 * encryption-as-a-service, and detailed audit logging. Unlike .env files:
 *   - Vault secrets can auto-expire (TTL-based) — leaked credentials become useless
 *   - Every access is logged to an audit backend (file, syslog, etc.)
 *   - Vault supports policies (e.g., "app" role can read /secret/data/app/*")
 *   - Dynamic database credentials mean each app instance gets its own DB user
 *
 * SECURITY NOTE ON VAULT TOKENS:
 * The VAULT_TOKEN env var here is a "bootstrap" token. In production you
 * should use Vault Agent or the Kubernetes/Auth method so tokens are
 * injected at runtime and never appear in environment variables at all.
 */

const vault = require("node-vault");
const winston = require("winston");

// ── Logger Configuration ──────────────────────────────────────────────────
// NEVER log secret values. Use [REDACTED] to indicate a value was retrieved.
const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

// ── Vault Client Initialization ───────────────────────────────────────────
let vaultClient;

/**
 * Creates and returns a configured Vault client.
 *
 * BOOTSTRAPPING FROM ENVIRONMENT:
 *   VAULT_ADDR  - URL of the Vault server (e.g., https://vault.example.com:8200)
 *   VAULT_TOKEN - Authentication token (in production, use Vault Agent instead)
 *   VAULT_NAMESPACE - Optional namespace for Vault Enterprise multi-tenancy
 *
 * SECURITY: The Vault address should use HTTPS. If VAULT_ADDR starts with
 * "http://" (not "https://"), we log a warning because secrets travel
 * unencrypted over the network.
 */
function getClient() {
  if (!vaultClient) {
    const vaultAddr = process.env.VAULT_ADDR;
    const vaultToken = process.env.VAULT_TOKEN;

    if (!vaultAddr) {
      throw new Error(
        "VAULT_ADDR environment variable is required. " +
        "Example: https://vault.example.com:8200"
      );
    }
    if (!vaultToken) {
      throw new Error(
        "VAULT_TOKEN environment variable is required. " +
        "In production, prefer Vault Agent or Kubernetes auth to avoid " +
        "placing tokens in environment variables."
      );
    }

    // Warn if not using TLS — secrets will be sent in plaintext
    if (vaultAddr.startsWith("http://")) {
      logger.warn("Vault address is NOT using HTTPS — secrets are unencrypted in transit", {
        vaultAddr,
      });
    }

    const config = {
      endpoint: vaultAddr,
      token: vaultToken,
      // Increase timeout for large secret payloads or slow networks
      "request-timeout": parseInt(process.env.VAULT_TIMEOUT || "30000", 10),
    };

    // Vault Enterprise namespaces allow multi-tenancy within a single Vault cluster
    if (process.env.VAULT_NAMESPACE) {
      config.namespace = process.env.VAULT_NAMESPACE;
      logger.info("Using Vault Enterprise namespace", {
        namespace: config.namespace,
      });
    }

    vaultClient = vault(config);
    logger.info("Vault client initialized", {
      vaultAddr,
      tokenPrefix: `${vaultToken.substring(0, 6)}...[REDACTED]`,
    });
  }
  return vaultClient;
}

/**
 * Fetches a KV v2 secret from Vault.
 *
 * Vault KV v2 stores secrets at the path: secret/data/<mount>/path
 * The actual secret values are nested under a "data.data" wrapper because
 * KV v2 tracks metadata (versions, custom metadata, deletion time) separately.
 *
 * @param {string} path - The secret path (e.g., "secret/data/app/database")
 * @returns {Promise<object>} The secret key-value pairs.
 */
async function getSecret(path) {
  const client = getClient();

  logger.info("Fetching secret from Vault", { path });

  try {
    const response = await client.read(path);

    // KV v2 engine wraps secret data under response.data.data
    // KV v1 engine has data directly under response.data
    const secretData = response.data?.data || response.data || {};

    logger.info("Secret retrieved successfully", {
      path,
      keyCount: Object.keys(secretData).length,
      value: "[REDACTED]",
      // KV v2 exposes version info — useful for debugging rotation
      version: response.data?.metadata?.version || "N/A",
    });

    return secretData;
  } catch (error) {
    if (error.message?.includes("permission denied") || error.statusCode === 403) {
      logger.error("Access denied to Vault secret — check Vault policies", {
        path,
        error: error.message,
      });
      throw new Error(`Permission denied for Vault path '${path}'`);
    }
    if (error.statusCode === 404) {
      logger.error("Secret path not found in Vault", { path });
      throw new Error(`Vault secret at '${path}' not found`);
    }
    logger.error("Failed to fetch secret from Vault", {
      path,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Fetches database credentials from Vault.
 *
 * DYNAMIC SECRETS: Vault can generate short-lived database credentials
 * on demand via its database secrets engine. Instead of reading a static
 * secret, you request a lease — Vault creates a new DB user, returns the
 * credentials, and automatically revokes them when the lease expires.
 *
 * @param {string} credsPath - Dynamic creds role path (e.g., "database/creds/app-readonly")
 * @returns {Promise<object>} Database credentials with lease info.
 */
async function getDatabaseCredentials(credsPath = "database/creds/app-readonly") {
  const client = getClient();

  logger.info("Requesting dynamic database credentials from Vault", {
    credsPath,
  });

  try {
    const response = await client.read(credsPath);
    const data = response.data;

    // Dynamic secrets include lease information — the credentials will
    // be automatically revoked after lease_duration seconds
    logger.info("Dynamic database credentials generated", {
      credsPath,
      username: data.username,
      password: "[REDACTED]",
      leaseId: data.lease_id ? `${data.lease_id.substring(0, 20)}...` : "N/A",
      leaseDuration: data.lease_duration,
      renewable: data.renewable || false,
    });

    return {
      username: data.username,
      password: data.password,
      leaseId: data.lease_id,
      leaseDuration: data.lease_duration,
      renewable: data.renewable || false,
    };
  } catch (error) {
    logger.error("Failed to generate dynamic database credentials", {
      credsPath,
      error: error.message,
    });
    throw new Error(`Vault database credentials error: ${error.message}`);
  }
}

/**
 * Fetches JWT configuration from Vault.
 *
 * @param {string} path - KV path to JWT config (e.g., "secret/data/app/jwt")
 * @returns {Promise<object>} JWT configuration.
 */
async function getJwtConfig(path = "secret/data/app/jwt") {
  const secret = await getSecret(path);

  if (!secret.secretKey) {
    logger.error("JWT secret missing secretKey field", { path });
    throw new Error(`Vault secret at '${path}' must contain a 'secretKey' field`);
  }

  logger.info("JWT configuration retrieved", {
    path,
    algorithm: secret.algorithm || "HS256",
    secretKey: "[REDACTED]",
  });

  return {
    algorithm: secret.algorithm || "HS256",
    secretKey: secret.secretKey,
    issuer: secret.issuer || "secrets-management-app",
    expiresIn: secret.expiresIn || "15m",
  };
}

/**
 * Fetches a generic API key from Vault.
 *
 * @param {string} path - KV path to the API key secret.
 * @param {string} keyField - Field name containing the key (default: "apiKey").
 * @returns {Promise<object>} Configuration object with the API key.
 */
async function getApiKey(path, keyField = "apiKey") {
  const secret = await getSecret(path);

  if (!secret[keyField]) {
    logger.error("API key secret missing required field", { path, keyField });
    throw new Error(`Secret at '${path}' must contain '${keyField}'`);
  }

  logger.info("API key retrieved", {
    path,
    keyField,
    apiKeyPrefix: `${secret[keyField].substring(0, 7)}...`,
  });

  return {
    [keyField]: secret[keyField],
    ...Object.fromEntries(
      Object.entries(secret).filter(([k]) => k !== keyField)
    ),
  };
}

module.exports = {
  getSecret,
  getDatabaseCredentials,
  getJwtConfig,
  getApiKey,
  getClient,
};

File: examples/python/vault_secrets.py
"""
HashiCorp Vault Integration Module
====================================
This module demonstrates how to securely fetch secrets from a HashiCorp
Vault KV v2 secret engine using the `hvac` Python client.

Security rationale
------------------
* The Vault address and authentication token are read from environment
  variables (VAULT_ADDR, VAULT_TOKEN).  In production you would use
  short-lived tokens issued by Vault's auth methods (AppRole, Kubernetes,
  JWT/OIDC) rather than long-lived root tokens.
* Vault's KV v2 engine versions every secret write, so you can roll back
  to a previous value if a rotation goes wrong.
* Dynamic secrets (e.g. database credentials generated on-the-fly) have
  a built-in lease / TTL, meaning credentials are automatically revoked
  after a configurable period — even if the application crashes.
* All secret values stay in memory and are never logged.
"""

import logging
import os
from typing import Any

import hvac
from hvac.exceptions import InvalidRequest, VaultError

# ---------------------------------------------------------------------------
# Structured logging — same pattern as aws_secrets.py.
# We log *which* secret was accessed and *whether* it succeeded, never the
# actual secret material.
# ---------------------------------------------------------------------------
logger = logging.getLogger("vault_secrets")
logger.setLevel(logging.INFO)


def _log_event(event: str, path: str | None = None, **extra: Any) -> None:
    """Emit a structured log entry, redacting any sensitive-looking keys."""
    payload = {"event": event, "vault_path": path, **extra}
    for k, v in payload.items():
        if any(s in k.lower() for s in ("key", "token", "secret", "password")):
            payload[k] = "***REDACTED***"
    logger.info(payload)


class VaultSecretsManager:
    """High-level wrapper around HashiCorp Vault KV v2 secrets.

    Vault KV v2 paths look like:  secret/data/<mount>/<name>
    The actual data lives under response["data"]["data"].
    Metadata (version, created time, custom metadata) lives under
    response["data"]["metadata"].
    """

    # Default paths — override via constructor args or env vars.
    DEFAULT_MOUNT = "secret"
    DEFAULT_OPENAI_PATH = "openai/config"
    DEFAULT_DB_PATH = "database/credentials"
    DEFAULT_JWT_PATH = "jwt/config"

    def __init__(
        self,
        vault_addr: str | None = None,
        token: str | None = None,
        mount: str | None = None,
        namespace: str | None = None,
        verify_tls: bool = True,
    ) -> None:
        # ------------------------------------------------------------------
        # Bootstrapping from environment variables:
        #   VAULT_ADDR  — URL of the Vault server (required)
        #   VAULT_TOKEN — client token for authentication
        #   VAULT_NAMESPACE — Vault enterprise namespace (optional)
        #
        # SECURITY: Tokens should be short-lived and obtained via an
        # auth method (AppRole, K8s, JWT/OIDC), NOT a root token.
        # ------------------------------------------------------------------
        self._addr = vault_addr or os.environ["VAULT_ADDR"]
        self._token = token or os.environ.get("VAULT_TOKEN", "")
        self._mount = mount or os.getenv("VAULT_KV_MOUNT", self.DEFAULT_MOUNT)
        self._namespace = namespace or os.getenv("VAULT_NAMESPACE")

        _log_event(
            "vault_client_init",
            extra={
                "addr": self._addr,  # URL is not secret
                "mount": self._mount,
                "has_token": bool(self._token),  # log presence, not value
            },
        )

        self._client = hvac.Client(
            url=self._addr,
            token=self._token,
            namespace=self._namespace,
            verify=verify_tls,
        )

        # Verify connectivity at construction time so we fail fast.
        if not self._client.is_authenticated():
            raise VaultError(
                "Vault client is not authenticated — check VAULT_ADDR and VAULT_TOKEN"
            )

        # Resolve secret paths
        self._openai_path = os.getenv(
            "VAULT_SECRET_OPENAI", self.DEFAULT_OPENAI_PATH
        )
        self._db_path = os.getenv(
            "VAULT_SECRET_DB", self.DEFAULT_DB_PATH
        )
        self._jwt_path = os.getenv(
            "VAULT_SECRET_JWT", self.DEFAULT_JWT_PATH
        )

    # ------------------------------------------------------------------
    # Core fetch
    # ------------------------------------------------------------------
    def get_secret(self, path: str, version: int | None = None) -> dict[str, Any]:
        """Read a secret from KV v2 and return its data dict.

        Args:
            path:    Secret path *relative to the mount* (e.g. "database/creds").
            version: Optional version number to fetch a specific version.

        In KV v2 the API is:  GET /v1/secret/data/<path>
        The hvac client handles the "secret/data/" prefix for us when
        using ``kv.v2.read_secret_version()``.

        Raises:
            InvalidRequest: secret does not exist or access is denied.
            VaultError:     generic Vault communication error.
        """
        full_path = f"{self._mount}/{path}"
        _log_event("fetching_secret", path=full_path, requested_version=version)

        try:
            read_kwargs: dict[str, Any] = {"path": path, "mount_point": self._mount}
            if version is not None:
                read_kwargs["version"] = version
            response = self._client.secrets.kv.v2.read_secret_version(**read_kwargs)
        except InvalidRequest as exc:
            _log_event("fetch_secret_denied", path=full_path)
            raise
        except VaultError as exc:
            _log_event("fetch_secret_error", path=full_path, error=str(exc))
            raise

        # KV v2 nests actual data under "data" → "data"
        data: dict[str, Any] = response["data"]["data"]
        metadata: dict[str, Any] = response["data"]["metadata"]

        _log_event(
            "secret_fetched",
            path=full_path,
            version=metadata.get("version"),
            created_time=metadata.get("created_time"),
        )
        return data

    # ------------------------------------------------------------------
    # Typed convenience methods
    # ------------------------------------------------------------------
    def get_database_credentials(self) -> dict[str, Any]:
        """Return database credentials from Vault.

        Expected secret shape:
        {
            "engine": "postgresql",
            "host": "db.example.com",
            "port": 5432,
            "username": "app_user",
            "password": "<sensitive>",
            "dbname": "app_db"
        }

        NOTE: For truly dynamic credentials, use Vault's database secrets
        engine which generates a new username/password pair with a lease
        TTL.  That approach is even more secure because credentials are
        automatically revoked when the lease expires.
        """
        creds = self.get_secret(self._db_path)
        required = {"host", "username", "password", "dbname"}
        missing = required - set(creds.keys())
        if missing:
            raise ValueError(f"DB secret missing fields: {missing}")
        return creds

    def get_jwt_config(self) -> dict[str, Any]:
        """Return JWT signing configuration from Vault.

        Expected shape: {"algorithm": "HS256", "secret_key": "...", ...}
        """
        config = self.get_secret(self._jwt_path)
        if "secret_key" not in config:
            raise ValueError("JWT secret missing 'secret_key' field")
        return config

    def get_openai_config(self) -> dict[str, Any]:
        """Return OpenAI API configuration from Vault.

        Expected shape: {"api_key": "sk-...", "organization_id": "org-...", ...}
        """
        config = self.get_secret(self._openai_path)
        if "api_key" not in config:
            raise ValueError("OpenAI secret missing 'api_key' field")
        return config

    # ------------------------------------------------------------------
    # Versioning helpers
    # ------------------------------------------------------------------
    def get_secret_versions(self, path: str) -> list[int]:
        """List all versions of a KV v2 secret.

        Vault keeps every version by default, which is invaluable for
        auditing and rollback.  This method returns a list of version
        numbers.
        """
        try:
            metadata = self._client.secrets.kv.v2.read_secret_metadata(
                path=path, mount_point=self._mount
            )
        except InvalidRequest:
            _log_event("list_versions_denied", path=path)
            raise

        versions = list(metadata["data"]["versions"].keys())
        _log_event("versions_listed", path=path, count=len(versions))
        return [int(v) for v in versions]

    def get_previous_version(self, path: str) -> dict[str, Any]:
        """Fetch the previous version of a secret (useful for rollback)."""
        versions = self.get_secret_versions(path)
        if len(versions) < 2:
            raise ValueError(f"Secret '{path}' has only one version — no previous")
        previous = versions[-2]
        return self.get_secret(path, version=previous)


# ---------------------------------------------------------------------------
# Quick-start demonstration
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")

    try:
        mgr = VaultSecretsManager()
        db = mgr.get_database_credentials()
        print(f"DB host: {db['host']}")
    except KeyError as exc:
        print(f"VAULT_ADDR or VAULT_TOKEN not set (expected in local dev): {exc}")
    except Exception as exc:
        print(f"Vault error (expected in local dev): {exc}")